Tools/yocto/README.md

```diff
@@ -0,0 +1,268 @@
+
+A trick to speed up the whole rebuild process is to store the yocto sstate
+and download caches outside of the WebKitBuild directory, so they can be
+reused between different builds.
+
+To enable that, add this lines to your ```~/.bashrc``` file
+
+```
+# Export shared directories for yocto caches
+export DL_DIR="${HOME}/.cache/yocto/downloads"
+export SSTATE_DIR="${HOME}/.cache/yocto/sstate"
+export BB_ENV_PASSTHROUGH_ADDITIONS="${BB_ENV_PASSTHROUGH_ADDITIONS} DL_DIR SSTATE_DIR"
+```
+
+### Disabling automatic wipe on change
+
+The script ```cross-toolchain-helper``` will automatically wipe and rebuild the
+cross-toolchain and images for the target board if it detects anything has changed
+on the target source definitions.
+
+To avoid this you can pass the flag ```--no-wipe-on-change``` to ```cross-toolchain-helper```
+or you can set the environment variable ```WEBKIT_CROSS_WIPE_ON_CHANGE=0```
+
+However it is not recommended that you do this, instead try to enable an external Yocto cache
+to have very fast rebuilds (see above).
+
+
+### Disabling export default environment
+
+The script ```cross-toolchain-helper``` will automatically export several environment variables
+defined in file ```${WebKitDirectory}/Tools/yocto/targets.conf``` for the given target.
+This environment variables are usually used to change the default build parameters for the
+script ```build-webkit```
+
+If you want to disable this you can pass the flag ```--no-export-default-environment``` to ```cross-toolchain-helper```
+or you can set the environment variable ```WEBKIT_CROSS_EXPORT_ENV=0```
+
+
+## Practical examples on how to use this
+
+This is an example on how to build and run WPE for development on the RPi:
+
+### Build the base image, flash the board and clone the WebKit repo
+
+1. Build the image that will be flashed on the board:
+
+```
+user@workstation $ Tools/Scripts/cross-toolchain-helper --cross-target=rpi3-32bits-mesa --build-image
+```
+
+2. At the end of the build it will print the path to the image, if you missed it
+   run the command again and it will print the path again (doesn't rebuild if not needed).
+   Example:
+
+```
+user@workstation $ Tools/Scripts/cross-toolchain-helper --cross-target=rpi3-32bits-mesa --build-image
+cross-toolchain-helper INFO: Image already built at: /home/igalia/webkit/WebKitBuild/CrossToolChains/rpi3-32bits-mesa/build/image/core-image-webkit-dev-ci-tools-tests.wic.bz2
+```
+
+3. Flash a SDCard with this image
+
+  * 3.1. Use bmaptool for a faster copy
+
+```
+# deb distro: sudo apt install bmap-tools
+# rpm distro: sudo dnf install bmap-tools
+user@workstation $ bmaptool copy /path/to/core-image-webkit-dev-ci-tools-tests.wic.bz2 /dev/mmcblk0
+```
+
+  * 3.2. Or alternatively use dd
+
+```
+user@workstation $ cat /path/to/core-image-webkit-dev-ci-tools-tests.wic.bz2 | bzip2 -dc | dd of=/dev/mmcblk0 status=progress bs=4k
+```
+
+4. Grow the rootfs partition to use all the available space on the SDcard
```

Comment:
Is this step needed also when the rootfs is flashed with bmaptool? I'm not sure.

Reply:
It is needed. As far as I can see bmaptool does not grow the partition to the space available on the block.
Tested:

1. Clean the sdcard (just in case)
```
$ sudo dd if=/dev/zero of=/dev/mmcblk0 bs=1M count=20
20+0 records in
20+0 records out
20971520 bytes (21 MB, 20 MiB) copied, 0.34829 s, 60.2 MB/s

$ sudo fdisk -l /dev/mmcblk0
Disk /dev/mmcblk0: 59 GiB, 63281561600 bytes, 123596800 sectors
Units: sectors of 1 * 512 = 512 bytes
Sector size (logical/physical): 512 bytes / 512 bytes
I/O size (minimum/optimal): 512 bytes / 512 bytes
```

2. Copy the image with bmaptool
```
$ sudo bmaptool copy webkit-dev-ci-tools.wic.bz2 /dev/mmcblk0
bmaptool: info: discovered bmap file 'webkit-dev-ci-tools.wic.bmap'
bmaptool: info: block map format version 2.0
bmaptool: info: 1069978 blocks of size 4096 (4.1 GiB), mapped 551797 blocks (2.1 GiB or 51.6%)
bmaptool: info: copying image 'webkit-dev-ci-tools.wic.bz2' to block device '/dev/mmcblk0' using bmap file 'webkit-dev-ci-tools.wic.bmap'
bmaptool: info: 100% copied
bmaptool: info: synchronizing '/dev/mmcblk0'
bmaptool: info: copying time: 39.7s, copying speed 54.2 MiB/sec
```
3. Show the partitions after bmaptool copy
```
$ sudo fdisk -l /dev/mmcblk0
Disk /dev/mmcblk0: 59 GiB, 63281561600 bytes, 123596800 sectors
Units: sectors of 1 * 512 = 512 bytes
Sector size (logical/physical): 512 bytes / 512 bytes
I/O size (minimum/optimal): 512 bytes / 512 bytes
Disklabel type: dos
Disk identifier: 0x657f5e7b

Device         Boot  Start     End Sectors Size Id Type
/dev/mmcblk0p1 *      8192  112677  104486  51M  c W95 FAT32 (LBA)
/dev/mmcblk0p2      114688 8559819 8445132   4G 83 Linux
```

As you see the last partition is only 4GBs (what is set on the image) instead of 60GB available on the sdcard.

If now I execute the other steps on the README then I get it to use all the space available.

```
$ sudo growpart /dev/mmcblk0 2
CHANGED: partition=2 start=114688 old: size=8445132 end=8559820 new: size=123482079,end=123596767

$ sudo resize2fs /dev/mmcblk0p2
resize2fs 1.44.5 (15-Dec-2018)
Resizing the filesystem on /dev/mmcblk0p2 to 15435259 (4k) blocks.
The filesystem on /dev/mmcblk0p2 is now 15435259 (4k) blocks long.


$ sudo fdisk -l /dev/mmcblk0
Disk /dev/mmcblk0: 59 GiB, 63281561600 bytes, 123596800 sectors
Units: sectors of 1 * 512 = 512 bytes
Sector size (logical/physical): 512 bytes / 512 bytes
I/O size (minimum/optimal): 512 bytes / 512 bytes
Disklabel type: dos
Disk identifier: 0x657f5e7b

Device         Boot  Start       End   Sectors  Size Id Type
/dev/mmcblk0p1 *      8192    112677    104486   51M  c W95 FAT32 (LBA)
/dev/mmcblk0p2      114688 123596766 123482079 58.9G 83 Linux
```